### ISSUE TITLE
[HOLD for payment 2022-12-09] [$1000] [Bug] mWeb - If you close the keyboard using native back button then it will keep reopening while selecting the contacts - reported by @thesahindia

### ISSUE THREAD

**Issue by mvtglobally** [Expensify/App#11094]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Tap on :green-plus-button: > New group
2. Close the keyboard by tapping native back button
3. Select a user

## Expected Result:
The keyboard should not open

## Actual Result:
The keyboard keeps opening

## Workaround:
unknown 

## Platform:

Where is this issue occurring?

- Mobile Web

**Version Number:** 1.1.99-0
**Reproducible in staging?:** Y 
**Reproducible in production?:** Y 
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation 

https://user-images.githubusercontent.com/43995119/190953813-2d52afbf-0724-4aff-b372-5570ef9f37a5.mov

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @thesahindia
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1661865140075919

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by RachCHopkins** [comment]
I'm still unable to reproduce this. Is this still happening @mvtglobally ?

**Comment by techievivek** [comment]
I am able to reproduce this on staging so we can definitely work on this one.

**Comment by Beamanator** [comment]
@techievivek just wondering why unassign both you and me? 🤔

**Comment by techievivek** [comment]
@Beamanator Oops my bad, I didn't mean to unassign you, I did that by mistake. 😞

**Comment by Beamanator** [comment]
Aah ok no worries :D

**Comment by Beamanator** [comment]
Hmm I remember hearing we're holding most of these keyboard-related issues for some upcoming "keyboard initiative" - asking in Slack about this before putting it on hold officially

**Comment by mananjadhav** [comment]
Yeah, we should hold them. This is the issue where we're tagging all Keyboard related issue.

https://github.com/Expensify/App/issues/10273

**Comment by Beamanator** [comment]
Cool good call, putting on hold

**Comment by Beamanator** [comment]
Still on hold

**Comment by tgolen** [comment]
As I was looking into this, I was struck by the feeling that this is intentional. Once you select a user, the focus is intentionally put back into the search box. This is what causes the keyboard to open. Because of that, I am going to close this one since it's working as intended I think. Feel free to reopen and plead your case!

**Comment by thesahindia** [comment]
> As I was looking into this, I was struck by the feeling that this is intentional. Once you select a user, the focus is intentionally put back into the search box. This is what causes the keyboard to open. Because of that, I am going to close this one since it's working as intended I think. Feel free to reopen and plead your case!

I think we should change this because it doesn't happen at the native app and it's also a problem when you are selecting multiple contacts for creating a group or splitting bill, it will be annoying for the user to keep closing the keyboard when they just wanna select first few contacts.

**Comment by tgolen** [comment]
Thank you for pointing out that it's inconsistent with native apps. I tested it out, and you're right. The native behavior is much better and I'd like to make the UX consistent across devices. Reopening!

**Comment by Beamanator** [comment]
Assigning @tgolen since you're the keyboard leader 👍

**Comment by tgolen** [comment]
OK, you can assign it to me for now, but I'm not going to be fixing this. My intention is to get [my PR](https://github.com/Expensify/App/pull/11586) merged which will remove the hold on this and allow a contributor to propose a fix. I'll be happy to look at proposals though!

**Comment by tgolen** [comment]
This is no longer on hold and can have proposals created for it with the new code.

**Comment by Christinadobrzyn** [comment]
Posted job in Upwork: 

Internal posting: https://www.upwork.com/ab/applicants/1587648449081278464/job-details
External posting: https://www.upwork.com/jobs/~0124bea2e5e1d56e2b

**Comment by JmillsExpensify** [comment]
@Christinadobrzyn Given that we're still early and now proposals have been accepted, I'm assigning myself per the new BZ process: https://stackoverflow.com/c/expensify/questions/14597/14598#14598.

**Comment by Someone-here** [comment]
### Proposal

I noticed that when selecting an option, The code below would focus the TextInput after selecting.

![selectRow](https://user-images.githubusercontent.com/77237602/200127379-bad0fce4-f506-47ba-8672-7662a12b76d5.png)

We can solve this by passing the `shouldFocusOnSelectRow` prop as false.

![props](https://user-images.githubusercontent.com/77237602/200127451-036f84ea-1e18-439b-baed-12b6df70f737.png)

Here is the result after doing that:


https://user-images.githubusercontent.com/77237602/200127564-d3e142b0-3970-488a-8a6e-348cd9ef0f38.mp4

**Comment by mananjadhav** [comment]
Thanks for the interest here @Someone-here. While I am trying to review your proposal, I don't understand why would this work on mobile App properly but not mWeb. Your proposal _might_ fix it, but do you have the cause why the behavior is different on mWeb only?

**Comment by Someone-here** [comment]
You can check the comment in the code:
Input is permanently focused on native platforms, so we always highlight the text inside of it.
So this is an asymmetry with native and web platforms

**Comment by mananjadhav** [comment]
> You can check the comment in the code:

@Someone-here but again the comment is for the `native` platforms and you're suggesting to set it as false, which would essentially remove the focus? I still feel we're not solving the root cause and I don't see the discrepancy in the web and native platforms in your suggestion/comments.

**Comment by JmillsExpensify** [comment]
Still waiting for approved proposals. I'm increasing the price to $500.

**Comment by mananjadhav** [comment]
Waiting for proposals.

**Comment by JmillsExpensify** [comment]
On that note, it's been a week and it's time to double price. We're still open for proposals! Now at $1,000.

**Comment by tgolen** [comment]
Still looking for proposals.

**Comment by JmillsExpensify** [comment]
Ideally we'll have Margelo or an internal engineer take this issue over soon. We're still struggling to get external proposals.

**Comment by s77rt** [comment]
### Proposal
```diff
diff --git a/src/components/OptionsSelector/BaseOptionsSelector.js b/src/components/OptionsSelector/BaseOptionsSelector.js
index 7fce40647..2563cb949 100755
--- a/src/components/OptionsSelector/BaseOptionsSelector.js
+++ b/src/components/OptionsSelector/BaseOptionsSelector.js
@@ -36,7 +36,8 @@ class BaseOptionsSelector extends Component {
         this.updateFocusedIndex = this.updateFocusedIndex.bind(this);
         this.scrollToIndex = this.scrollToIndex.bind(this);
         this.selectRow = this.selectRow.bind(this);
-        this.relatedTarget = null;
+        this.isFocused = false;
+        this.isScheduledToFocusOnSelectRow = false;
 
         const allOptions = this.flattenSections();
         this.state = {
@@ -206,12 +207,13 @@ class BaseOptionsSelector extends Component {
      */
     selectRow(option, ref) {
         if (this.props.shouldFocusOnSelectRow) {
-            // Input is permanently focused on native platforms, so we always highlight the text inside of it
-            setSelection(this.textInput, 0, this.props.value.length);
-            if (this.relatedTarget && ref === this.relatedTarget) {
+            if (this.isScheduledToFocusOnSelectRow) {
                 this.textInput.focus();
             }
-            this.relatedTarget = null;
+            this.isScheduledToFocusOnSelectRow = false;
+            if (this.isFocused) {
+                setSelection(this.textInput, 0, this.props.value.length);
+            }
         }
         this.props.onSelectRow(option);
 
@@ -239,11 +241,17 @@ class BaseOptionsSelector extends Component {
                 label={this.props.textInputLabel}
                 onChangeText={this.props.onChangeText}
                 placeholder={this.props.placeholderText || this.props.translate('optionsSelector.nameEmailOrPhoneNumber')}
+                onFocus={() => {
+                    this.isFocused = true;
+                }}
                 onBlur={(e) => {
+                    this.isFocused = false;
                     if (!this.props.shouldFocusOnSelectRow) {
                         return;
                     }
-                    this.relatedTarget = e.relatedTarget;
+                    if (e.relatedTarget) {
+                        this.isScheduledToFocusOnSelectRow = true;
+                    }
                 }}
                 selectTextOnFocus
                 blurOnSubmit={Boolean(this.state.allOptions.length)}
```

### Details
The issue can be divided into two problems;
1. This text is being selected even if the input is not focused. Setting text selection automatically make the textinput focused.
2. Checking whether the textinput should be focused again is broken
    1. On native it does nothing, but we are getting the expected behaviour because that's how the native app works (e.g. clicking on a row won't blur the text input); you can test by changing the value of shouldFocusOnSelectRow
    2. On web it's pretty much okay, just the use of ref on selectRow is redundant I got it removed and the issue here is actually the text selection (issue 1)

I fixed (1) by moving the selection at last and only after checking that the input is focused.
I fixed (2) by refactoring the code


https://user-images.githubusercontent.com/16493223/202818789-f102e813-0018-406a-b477-fb3195aa47b5.mp4

**Comment by s77rt** [comment]
### Proposal (Updated)
```diff
diff --git a/src/components/OptionsSelector/BaseOptionsSelector.js b/src/components/OptionsSelector/BaseOptionsSelector.js
index 7fce40647..26c0b4983 100755
--- a/src/components/OptionsSelector/BaseOptionsSelector.js
+++ b/src/components/OptionsSelector/BaseOptionsSelector.js
@@ -36,7 +36,8 @@ class BaseOptionsSelector extends Component {
         this.updateFocusedIndex = this.updateFocusedIndex.bind(this);
         this.scrollToIndex = this.scrollToIndex.bind(this);
         this.selectRow = this.selectRow.bind(this);
-        this.relatedTarget = null;
+        this.isFocused = false;
+        this.isScheduledToFocusOnSelectRow = false;
 
         const allOptions = this.flattenSections();
         this.state = {
@@ -206,12 +207,13 @@ class BaseOptionsSelector extends Component {
      */
     selectRow(option, ref) {
         if (this.props.shouldFocusOnSelectRow) {
-            // Input is permanently focused on native platforms, so we always highlight the text inside of it
-            setSelection(this.textInput, 0, this.props.value.length);
-            if (this.relatedTarget && ref === this.relatedTarget) {
+            if (this.isScheduledToFocusOnSelectRow) {
                 this.textInput.focus();
+                this.isScheduledToFocusOnSelectRow = false;
+            }
+            if (this.isFocused) {
+                setSelection(this.textInput, 0, this.props.value.length);
             }
-            this.relatedTarget = null;
         }
         this.props.onSelectRow(option);
 
@@ -239,11 +241,14 @@ class BaseOptionsSelector extends Component {
                 label={this.props.textInputLabel}
                 onChangeText={this.props.onChangeText}
                 placeholder={this.props.placeholderText || this.props.translate('optionsSelector.nameEmailOrPhoneNumber')}
+                onFocus={() => {
+                    this.isFocused = true;
+                }}
                 onBlur={(e) => {
-                    if (!this.props.shouldFocusOnSelectRow) {
-                        return;
+                    this.isFocused = false;
+                    if (this.props.shouldFocusOnSelectRow && e.relatedTarget) {
+                        this.isScheduledToFocusOnSelectRow = true;
                     }
-                    this.relatedTarget = e.relatedTarget;
                 }}
                 selectTextOnFocus
                 blurOnSubmit={Boolean(this.state.allOptions.length)}
```

### Update
Made the code clearer

**Comment by mananjadhav** [comment]
@s77rt Thanks for the proposal. The issue description mentions mWeb with the Android emulator. I applied your proposal, but the issue persists with Android Chrome.  Can you confirm if it works on Android chrome too?

Also to an extent I can see the usage for `isFocused`, but not sure why we would remove `relatedTarget`?

**Comment by s77rt** [comment]
### Proposal (Updated)
```diff
diff --git a/src/components/OptionsSelector/BaseOptionsSelector.js b/src/components/OptionsSelector/BaseOptionsSelector.js
index 7fce40647..9d210d08a 100755
--- a/src/components/OptionsSelector/BaseOptionsSelector.js
+++ b/src/components/OptionsSelector/BaseOptionsSelector.js
@@ -8,7 +8,10 @@ import FixedFooter from '../FixedFooter';
 import OptionsList from '../OptionsList';
 import CONST from '../../CONST';
 import styles from '../../styles/styles';
+import compose from '../../libs/compose';
 import withLocalize, {withLocalizePropTypes} from '../withLocalize';
+import withWindowDimensions, {windowDimensionsPropTypes} from '../withWindowDimensions';
+import withKeyboardState from '../withKeyboardState';
 import TextInput from '../TextInput';
 import ArrowKeyFocusManager from '../ArrowKeyFocusManager';
 import KeyboardShortcut from '../../libs/KeyboardShortcut';
@@ -22,6 +25,7 @@ const propTypes = {
 
     ...optionsSelectorPropTypes,
     ...withLocalizePropTypes,
+    ...windowDimensionsPropTypes,
 };
 
 const defaultProps = {
@@ -36,6 +40,7 @@ class BaseOptionsSelector extends Component {
         this.updateFocusedIndex = this.updateFocusedIndex.bind(this);
         this.scrollToIndex = this.scrollToIndex.bind(this);
         this.selectRow = this.selectRow.bind(this);
+        this.isFocused = false;
         this.relatedTarget = null;
 
         const allOptions = this.flattenSections();
@@ -96,6 +101,20 @@ class BaseOptionsSelector extends Component {
     }
 
     componentDidUpdate(prevProps) {
+        if ((prevProps.isShown === true && this.props.isShown === false) || (prevProps.windowHeight < this.props.windowHeight)) {
+            // Blur input if keyboard got closed
+            // on Android the input does not lose focus on keyboard close
+            // See more https://github.com/facebook/react-native/issues/33532
+
+            // Blur input if window height got changed, same reason as above
+            // but specifically to cover mWeb since the keyboard event does not apply on mWeb
+
+            if (!this.textInput) {
+                return;
+            }
+            this.textInput.blur();
+        }
+
         if (_.isEqual(this.props.sections, prevProps.sections)) {
             return;
         }
@@ -206,12 +225,13 @@ class BaseOptionsSelector extends Component {
      */
     selectRow(option, ref) {
         if (this.props.shouldFocusOnSelectRow) {
-            // Input is permanently focused on native platforms, so we always highlight the text inside of it
-            setSelection(this.textInput, 0, this.props.value.length);
             if (this.relatedTarget && ref === this.relatedTarget) {
                 this.textInput.focus();
+                this.relatedTarget = null;
+            }
+            if (this.isFocused) {
+                setSelection(this.textInput, 0, this.props.value.length);
             }
-            this.relatedTarget = null;
         }
         this.props.onSelectRow(option);
 
@@ -239,11 +259,14 @@ class BaseOptionsSelector extends Component {
                 label={this.props.textInputLabel}
                 onChangeText={this.props.onChangeText}
                 placeholder={this.props.placeholderText || this.props.translate('optionsSelector.nameEmailOrPhoneNumber')}
+                onFocus={() => {
+                    this.isFocused = true;
+                }}
                 onBlur={(e) => {
-                    if (!this.props.shouldFocusOnSelectRow) {
-                        return;
+                    this.isFocused = false;
+                    if (this.props.shouldFocusOnSelectRow) {
+                        this.relatedTarget = e.relatedTarget;
                     }
-                    this.relatedTarget = e.relatedTarget;
                 }}
                 selectTextOnFocus
                 blurOnSubmit={Boolean(this.state.allOptions.length)}
@@ -318,4 +341,8 @@ class BaseOptionsSelector extends Component {
 BaseOptionsSelector.defaultProps = defaultProps;
 BaseOptionsSelector.propTypes = propTypes;
 
-export default withLocalize(BaseOptionsSelector);
+export default compose(
+    withLocalize,
+    withWindowDimensions,
+    withKeyboardState,
+)(BaseOptionsSelector);
```

### Update
Building on the previous proposal; same approach, just few points has been changed:
1. I had to use `ref` as they were before, because the `relatedTarget` will always be set (as far as I tested on mWeb - Chrome). On Safari this may be null on blur. On chrome will fallback to `<body>`.
2. On Android native, on keyboard dismiss, the text input does not lose focus (issue reported [here](https://github.com/facebook/react-native/issues/33532)) to make it work I'm using the HOC `withKeyboardState` to blur the input on keyboard close.
3. On Android web, same issue as above but we can't use `withKeyboardState` as it does not support web, instead I'm using `withWindowDimensions` to detect the change in the window height


This should work as expected on all platforms.


https://user-images.githubusercontent.com/16493223/202873692-7b97260b-4ef8-44c3-9c8c-9082a9668458.mp4



___
@mananjadhav Sorry about the previous proposal; my bad.

**Comment by mananjadhav** [comment]
Thanks for the updated proposal @s77rt.  

A couple of questions:
- How reliable is it to use `windowHeight` to detect keyboard change? I can see we're using a similar logic for PDFView. But I haven't first hand tried/tested this
- Is this issue on native too? Platforms mention only mobile web, so why are we adding additional handling for native?
- I still don't get why are we changing this behavior `// Input is permanently focused on native platforms, so we always highlight the text inside of it` ?

@tgolen do you think this is something we should handle in the `KeyboardAvoidingView` component? or outside the component?

**Comment by s77rt** [comment]
@mananjadhav 
> How reliable is it to use `windowHeight` to detect keyboard change? I can see we're using a similar logic for PDFView. But I haven't first hand tried/tested this

On Android, the window height will always change on keyboard open/close. Pretty reliable.

> Is this issue on native too? Platforms mention only mobile web, so why are we adding additional handling for native?

Yes and No; Yes: the input does not lose focus on keyboard close. No: the user can select another row without the keyboard getting open again (Android Only. iOS is fine). We are handling this to match the behaviour of Android with iOS; On keyboard close, blur the input.

> I still don't get why are we changing this behavior `// Input is permanently focused on native platforms, so we always highlight the text inside of it` ?

This is not true; only on Android it's "permanently" focused but that's more like a bug. On iOS the input is not "permanently" focused at all. (Your question is not much clear, if this does not answer your question let me know in details what you meant.)

**Comment by tgolen** [comment]
> @tgolen do you think this is something we should handle in the KeyboardAvoidingView component? or outside the component?

I am not following this issue too closely, so it's hard for me to say. My gut says "outside". However, the proposals in this issue seem really complex and come across to me as workarounds on top of workarounds.

I would like us to maybe think about trying to make this as simple as we can. If that means fixing bugs related to focus in upstream repos, then I think that's the better approach (not sure if that was mentioned yet).

We should really strive to not have different solutions for different platforms and instead, I'd like us to have a single solution that works on all platforms at almost any cost.

Lastly, I'll say there was a proposal in [this issue](https://github.com/Expensify/App/issues/10273) to use `useAnimatedKeyboard` for detecting when the keyboard is open/closed and creating our own KAV based on that concept. You can see in my latest comment there though that I am still not sure we are at that point yet.

Let's keep pushing proposal on this issue to try and find the MOST SIMPLE solution possible that works on all platforms.

**Comment by s77rt** [comment]
@tgolen If my proposal seems complex please ignore the code in `componentDidUpdate` for now.
The rest of the code fix the issue.

But..., Android is excluded (native and mWeb) because simply Android does not blur the input on keyboard close ([issue](https://github.com/facebook/react-native/issues/33532)).
That's where the `componentDidUpdate` comes to play (that's the workaround part) where we manually blur the input if we noticed that the keyboard has been closed; using the `isShown` prop for native and the `windowHeight` for mWeb

**Comment by tgolen** [comment]
OK, thank you.

I'm still not quite understanding the Android piece. I recently worked on input blur/focus and keyboard show/dismiss on the login form (in [this PR](https://github.com/Expensify/App/pull/12848)) and I found that the text input did lose focus when the keyboard was dismissed. Is this only related to specific Android devices (and if so, are they the ones we want to try and support)?

```
(prevProps.windowHeight < this.props.windowHeight)
```

Should this be added instead to `withKeyboardState` so that the HOC always supports mweb? Is this really the proper way of doing it? In my PR, it seemed like `withKeyboardState` was behaving properly for mWeb (both chrome and safari), but maybe I missed something.

I think we're on the right track for this proposal, it's just clearing up these remaining points. @mananjadhav feel free to jump in here again as well so that both of us are OK with the proposal.

**Comment by s77rt** [comment]
@tgolen

The blur problem as far as I know is Android related not device related.
Just to be clear, we want the input to lose focus on keyboard dismiss.

> Should this be added instead to withKeyboardState so that the HOC always supports mweb? 
No, the truth is we can't reliably support keyboard state on web, end of the story.
`withKeyboardState` does not support web as far as I know, it only seems to you working because the browser was doing the right thing.

Let me summarise:
1. You see the code that I proposed above, except the `componentDidUpdate` part, that's the fix. let's focus on that part first.
2. The code in `componentDidUpdate` is to support Android. On Android we have to manually blur the input on keyboard dismiss. We detect keyboard dismiss using the `isShown` prop (`withKeyboardState`) on native and the `windowHeight` for web. On android we can rely on the `windowHeight`, other platforms not really the case that's why we can't include this in `withKeyboardState`.



PS: Please tag me next time so I don't miss your messages.

**Comment by tgolen** [comment]
@s77rt Thanks for walking me through that. I'm generally fine with everything outside of `componentDidUpdate`, so be assured with that.

Regarding the stuff inside `componentDidUpdate`, we should move anything that only applies to android into an android specific file (eg. `index.android.js`) to match our cross-platform patterns. Then we should only have stuff that applies to web in a web-specific file (eg. `index.web.js`). 

Finally, I don't see how putting `(prevProps.windowHeight < this.props.windowHeight)` into `withKeyboardState` is any less reliable than putting that same logic in  `BaseOptionsSelector`. You're going to run into the same issues regardless, right? Wouldn't moving the logic to `withKeyboardState` make `withKeyboardState` more reliable now because it will provide support for mweb? 

In other words, if you add the logic to `BaseOptionsSelector` to support mWeb, then why not help the entire rest of the codebase support mWeb better by moving it to `withKeyboardState`?

Another way to look at it is that if that logic can't make `withKeyboardState` support mWeb, then how is it possible to allow `BaseOptionsSelector` to support mWeb?

Sorry to say it a few different ways, but maybe we are just on a slightly different wavelength and I'm trying to clear up any confusion in what I'm asking.

**Comment by s77rt** [comment]
@tgolen I know what you are saying, the thing is I'm worried if this would create bugs in the future. The window height is an indicator only, it's for the window height not for the keyboard state, but we use it as an indicator as the last resort and only where it's absolutely necessary like the case here. The window height is not reliable for mWeb - Safari or more specific on mWeb iOS. It just happens that the issue here does not involve iOS.

Ideally we want to support keyboard state on web, but unfortunately no browser provide an API for this.

**Comment by tgolen** [comment]
OK, I'm fine with that. Thanks for taking the time to understand what my concerns were and where I was coming from!

Since that's where you are landing, then I would ask for you to remove `(prevProps.windowHeight < this.props.windowHeight)` from your proposal. If we can't rely on it everywhere, then we shouldn't rely on it in just one spot. It leads to inconsistent behavior and code and that leads to maintainability and quality issues down the road.

**Comment by s77rt** [comment]
@tgolen That would keep Android mWeb unsupported

**Comment by tgolen** [comment]
Yep, I understand that. I'd rather have android mWeb unsupported then adding small workarounds like this in our code. If you prefer not to work on this, I can take it over (no shame in that!).

**Comment by s77rt** [comment]
I can work on it, just wanted to make sure you know.

**Comment by tgolen** [comment]
Great, thanks! Could I see a final proposal? I'd also still like to get approval from @eVoloshchak on the final.

**Comment by s77rt** [comment]
### Proposal (Updated)
```diff
diff --git a/src/components/OptionsSelector/BaseOptionsSelector.js b/src/components/OptionsSelector/BaseOptionsSelector.js
index 7fce40647..79393e755 100755
--- a/src/components/OptionsSelector/BaseOptionsSelector.js
+++ b/src/components/OptionsSelector/BaseOptionsSelector.js
@@ -8,7 +8,9 @@ import FixedFooter from '../FixedFooter';
 import OptionsList from '../OptionsList';
 import CONST from '../../CONST';
 import styles from '../../styles/styles';
+import compose from '../../libs/compose';
 import withLocalize, {withLocalizePropTypes} from '../withLocalize';
+import withKeyboardState from '../withKeyboardState';
 import TextInput from '../TextInput';
 import ArrowKeyFocusManager from '../ArrowKeyFocusManager';
 import KeyboardShortcut from '../../libs/KeyboardShortcut';
@@ -36,6 +38,7 @@ class BaseOptionsSelector extends Component {
         this.updateFocusedIndex = this.updateFocusedIndex.bind(this);
         this.scrollToIndex = this.scrollToIndex.bind(this);
         this.selectRow = this.selectRow.bind(this);
+        this.isFocused = false;
         this.relatedTarget = null;
 
         const allOptions = this.flattenSections();
@@ -96,6 +99,17 @@ class BaseOptionsSelector extends Component {
     }
 
     componentDidUpdate(prevProps) {
+        if (prevProps.isShown === true && this.props.isShown === false) {
+            // Blur input if keyboard got closed
+            // on Android the input does not lose focus on keyboard close
+            // See more https://github.com/facebook/react-native/issues/33532
+
+            if (!this.textInput) {
+                return;
+            }
+            this.textInput.blur();
+        }
+
         if (_.isEqual(this.props.sections, prevProps.sections)) {
             return;
         }
@@ -206,12 +220,13 @@ class BaseOptionsSelector extends Component {
      */
     selectRow(option, ref) {
         if (this.props.shouldFocusOnSelectRow) {
-            // Input is permanently focused on native platforms, so we always highlight the text inside of it
-            setSelection(this.textInput, 0, this.props.value.length);
             if (this.relatedTarget && ref === this.relatedTarget) {
                 this.textInput.focus();
+                this.relatedTarget = null;
+            }
+            if (this.isFocused) {
+                setSelection(this.textInput, 0, this.props.value.length);
             }
-            this.relatedTarget = null;
         }
         this.props.onSelectRow(option);
 
@@ -239,11 +254,14 @@ class BaseOptionsSelector extends Component {
                 label={this.props.textInputLabel}
                 onChangeText={this.props.onChangeText}
                 placeholder={this.props.placeholderText || this.props.translate('optionsSelector.nameEmailOrPhoneNumber')}
+                onFocus={() => {
+                    this.isFocused = true;
+                }}
                 onBlur={(e) => {
-                    if (!this.props.shouldFocusOnSelectRow) {
-                        return;
+                    this.isFocused = false;
+                    if (this.props.shouldFocusOnSelectRow) {
+                        this.relatedTarget = e.relatedTarget;
                     }
-                    this.relatedTarget = e.relatedTarget;
                 }}
                 selectTextOnFocus
                 blurOnSubmit={Boolean(this.state.allOptions.length)}
@@ -318,4 +336,7 @@ class BaseOptionsSelector extends Component {
 BaseOptionsSelector.defaultProps = defaultProps;
 BaseOptionsSelector.propTypes = propTypes;
 
-export default withLocalize(BaseOptionsSelector);
+export default compose(
+    withLocalize,
+    withKeyboardState,
+)(BaseOptionsSelector);
```

### Update
Removed `prevProps.windowHeight < this.props.windowHeight`

**Comment by tgolen** [comment]
Thanks! Could you please explain how you will move the Android stuff to its own specific android file too? I don't think it's necessary to show it in the code, but I'd like to at least know what plan you have for it.

**Comment by s77rt** [comment]
@tgolen 
Rename BaseOptionsSelector.js -> BaseOptionsSelector/index.js - contains the proposal without the android code
Add BaseOptionsSelector/index.android.js - contains the proposal with the android code
Move `propTypes` and `defaultProps` to BaseOptionsSelector/baseOptionsSelectorPropTypes.js

**Comment by aldo-expensify** [comment]
@s77rt about the proposal https://github.com/Expensify/App/issues/11094#issuecomment-1324355638, is there an advantage on keeping the focus state in `this.isFocused`? an alternative is to simple use `this.textInput.isFocused()` when you need and then you don't have to keep this value synchronized. I'm not sure if what I'm suggesting works in native or breaks.

**Comment by s77rt** [comment]
@aldo-expensify There is no advantage. If your suggestion works then we should use it instead of syncing the focus state using `onFocus` and `onBlur`. This is a prop on [react-native TextInput](https://reactnative.dev/docs/textinput#isfocused) so I'm more worried about web actually.
Nice catch

**Comment by mananjadhav** [comment]
Went through all the conversations. Looks like @s77rt's proposal will work. I tested the mWeb part of it and it worked fine.

I would've appreciated some better details [here](https://github.com/Expensify/App/issues/11094#issuecomment-1324622487) for Android specific code, but I am fine with that. We can work out those kinks on the PR.

**Comment by JmillsExpensify** [comment]
Ok great, so at this point we're ready to hire @s77rt right? Please assign the contributor when that next step is decided.

**Comment by s77rt** [comment]
### Proposal (Updated)
```diff
diff --git a/src/components/OptionsSelector/BaseOptionsSelector.js b/src/components/OptionsSelector/BaseOptionsSelector.js
index 7fce40647c..3210e86079 100755
--- a/src/components/OptionsSelector/BaseOptionsSelector.js
+++ b/src/components/OptionsSelector/BaseOptionsSelector.js
@@ -206,12 +206,13 @@ class BaseOptionsSelector extends Component {
      */
     selectRow(option, ref) {
         if (this.props.shouldFocusOnSelectRow) {
-            // Input is permanently focused on native platforms, so we always highlight the text inside of it
-            setSelection(this.textInput, 0, this.props.value.length);
             if (this.relatedTarget && ref === this.relatedTarget) {
                 this.textInput.focus();
+                this.relatedTarget = null;
+            }
+            if (this.textInput.isFocused()) {
+                setSelection(this.textInput, 0, this.props.value.length);
             }
-            this.relatedTarget = null;
         }
         this.props.onSelectRow(option);
```

### Update
- Removed Android native specific code. Android native does not suffer from the issue.
- Applied @aldo-expensify suggestion by using `isFocused()` method instead of a synced state variable

The change is much clearer now and does the job.


https://user-images.githubusercontent.com/16493223/204110237-4eadf680-137f-45c2-93c7-99e90da3603e.mp4

**Comment by mananjadhav** [comment]
Updated proposal [here](https://github.com/Expensify/App/issues/11094#issuecomment-1328120448) looks cleaner

**Comment by tgolen** [comment]
Great, that's looking really good now. Thanks @s77rt for working on that very diligently. 🟢  to move forward with being hired and creating a PR

**Comment by JmillsExpensify** [comment]
Awesome! @mananjadhav get we hire @s77rt? Let's get this PR merged in three days and quality for the 50% bonus.

**Comment by mananjadhav** [comment]
Yes @JmillsExpensify. I approved the proposal [here](https://github.com/Expensify/App/issues/11094?notification_referrer_id=NT_kwDOAC7UibI0NDI3MjY1ODAxOjMwNjkwNjU&notifications_query=is%3Aunread#issuecomment-1326485608) and commented on the updated proposal [here](https://github.com/Expensify/App/issues/11094?notification_referrer_id=NT_kwDOAC7UibI0NDI3MjY1ODAxOjMwNjkwNjU&notifications_query=is%3Aunread#issuecomment-1328182015)

All good from my side!

**Comment by JmillsExpensify** [comment]
Nice, thank you! I thought you could assign, though clearly I'm wrong about that. T minus 3 days to get this merged!

**Comment by s77rt** [comment]
Applied in [Upwork](https://www.upwork.com/jobs/~0124bea2e5e1d56e2b) (50% bonus)
[PR](https://github.com/Expensify/App/pull/13101) is ready

**Comment by JmillsExpensify** [comment]
@mananjadhav @tgolen PR is ready for ya'll. We still need to get the PR merged for the bonus to apply, though we're on the right track!

**Comment by mananjadhav** [comment]
PR is reviewed and approved. Just left to merge.

**Comment by JmillsExpensify** [comment]
Awesome, it looks like we're just waiting on someone to press merge. Who is that going to be?

**Comment by mananjadhav** [comment]
> Awesome, it looks like we're just waiting on someone to press merge. Who is that going to be?

@tgolen Will merge the PR.

**Comment by JmillsExpensify** [comment]
And, the PR is merged!

**Comment by JmillsExpensify** [comment]
Alright, the PR is merged, so let's circle back on payments. We still need to wait for the regression period to start/end, though this is where we landed:

 - Contributor: @s77rt $1,500 (incl. the 50% bonus)
 - Contributor+: @mananjadhav $1,500 (incl. the 50% bonus)
 - Issuer reporter: @thesahindia $250

**Comment by JmillsExpensify** [comment]
All contributors hired in Upwork. @tgolen @mananjadhav Can you confirm we're all good on the first think items in the checklist above?

**Comment by mananjadhav** [comment]
@JmillsExpensify I was trying to check the offending PR, but couldn't pinpoint to a proper one. I think this wasn't working since the beginning. How do we link offending PRs in such cases?

**Comment by JmillsExpensify** [comment]
That's a great question, though I will defer to @tgolen on that one.

**Comment by tgolen** [comment]
Yeah, I assume there could be situations where the bug existed ever since the code was initially written. In that case, there would still be a source PR for it, right? It would just be the very first PR that introduced the code? Though, I could still see a case where that might be nearly impossible to track down as well.

I think if we give a reasonable effort to find an offending PR, and we come up empty, then an exception is probably fine in those cases to not find a source of the regression.

**Comment by mananjadhav** [comment]
I'll try to take another stab at it over the weekend. Will post an update here.

--

@JmillsExpensify this is ready for payout with a 50% bonus.

**Comment by JmillsExpensify** [comment]
All paid out. 

> Alright, the PR is merged, so let's circle back on payments. We still need to wait for the regression period to start/end, though this is where we landed:
> 
> * Contributor: @s77rt $1,500 (incl. the 50% bonus)
> * Contributor+: @mananjadhav $1,500 (incl. the 50% bonus)
> * Issuer reporter: @thesahindia $250
>

**Comment by JmillsExpensify** [comment]
@tgolen @mananjadhav can both of you please fill out the BZ checklist above, and when you do, close this issue out? The regression test is a work in progress based on a larger keyboard regression test we'll do across various KAV cases, so that's done for now.

**Comment by tgolen** [comment]
> I'll try to take another stab at it over the weekend. Will post an update here.

@mananjadhav do you have an update on this?

**Comment by tgolen** [comment]
Bump @mananjadhav 

If no response today, I'll just mark that we couldn't locate the specific regression.

**Comment by mananjadhav** [comment]
I tried checking but the code has gone through multiple refactors, and I am unable to pinpoint as the component used different methods before for `setSelection`, `relatedTarget` handling.

https://github.com/Expensify/App/pull/7530
https://github.com/Expensify/App/pull/11673

But I am unable to build and check if it worked properly by reverting these fixes.

**Comment by tgolen** [comment]
OK, thank you! I'll just put down that it was inconclusive and we couldn't really find a direct source.
